<a href="https://colab.research.google.com/github/Ellie1220/Linear-Optimization/blob/main/Assignment_1_Linear_Optimization_ISOM_839_aichenglu.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment 1: Linear Optimization

In [ ]:
%pip install gurobipy

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.4/13.4 MB 46.8 MB/s eta 0:00:00


### Exercise 7.7 from the textbook (Data Models and Decisions)

In [ ]:
# Your answer: Modeling and solving using Gurobi Python
from itertools import product
from math import sqrt
import gurobipy as gp
from gurobipy import GRB

In [ ]:
earnings={'W':35,'X':47,'Y':60,'Z':140}
supplies = {'Aluminum': 600, 'Copper': 400, 'Magnesium': 800}
proportions = {'W': {'Aluminum': 0.30, 'Copper': 0.30, 'Magnesium': 0.40},
        'X': {'Aluminum': 0.40, 'Copper': 0.10, 'Magnesium': 0.50},
        'Y': {'Aluminum': 0.10, 'Copper': 0.25, 'Magnesium': 0.65},
        'Z': {'Aluminum': 0.15, 'Copper': 0.40, 'Magnesium': 0.45}}

In [ ]:
model = gp.Model("AlloyOptimization")


production = {}
for alloy in earnings.keys():
    production[alloy] = model.addVar(name=f"{alloy}_production")

# Objective function
model.setObjective(gp.quicksum(earnings[alloy] * production[alloy] for alloy in earnings.keys()), GRB.MAXIMIZE)

# Supply constraints
for metal in supplies.keys():
    model.addConstr(gp.quicksum(proportions[alloy][metal] * production[alloy] for alloy in earnings.keys()) <= supplies[metal],
                     name=f"{metal}_supply")

# Optimize the model
model.optimize()

# Print the optimal solution
if model.status == GRB.OPTIMAL:
    print("Optimal solution")
    for alloy in earnings.keys():
        print(f"Amount of alloy {alloy}: {production[alloy].x} tons")
    print(f"Total earnings: ${model.objVal}")
else:
    print("Optimization did not converge.")

Gurobi Optimizer version 11.0.0 build v11.0.0rc2 (linux64 - "Ubuntu 22.04.3 LTS")

CPU model: Intel(R) Xeon(R) CPU @ 2.20GHz, instruction set [SSE2|AVX|AVX2]
Thread count: 1 physical cores, 2 logical processors, using up to 2 threads

Optimize a model with 3 rows, 4 columns and 12 nonzeros
Model fingerprint: 0xa5b1691c
Coefficient statistics:
  Matrix range     [1e-01, 7e-01]
  Objective range  [4e+01, 1e+02]
  Bounds range     [0e+00, 0e+00]
  RHS range        [4e+02, 8e+02]
Presolve time: 0.01s
Presolved: 3 rows, 4 columns, 12 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    2.8200000e+32   6.000000e+30   2.820000e+02      0s
       3    1.5083871e+05   0.000000e+00   0.000000e+00      0s

Solved in 3 iterations and 0.02 seconds (0.00 work units)
Optimal objective  1.508387097e+05
Optimal solution
Amount of alloy W: 0.0 tons
Amount of alloy X: 903.2258064516129 tons
Amount of alloy Y: 0.0 tons
Amount of alloy Z: 774.1935483870967 tons
Total earnin

What actions would you recommend based on the optimal solution and the shadow price information?

Amount of alloy W: 0.0 tons

Amount of alloy X: 903.2258064516129 tons

Amount of alloy Y: 0.0 tons

Amount of alloy Z: 774.1935483870967 tons

Total earnings: $150838.70967741936

### Exercise 7.11 from the textbook (Data Models and Decisions)

In [ ]:
import gurobipy as gp
from gurobipy import GRB

# Data
plants = ['Amherst', 'Providence']
customers = [1, 2, 3, 4, 5]
production_costs = {'Amherst': [5, 3, 2, 6, 6],
                    'Providence': [7, 3, 3, 6, 7]}
transportation_costs = {'Amherst': [2, 3, 2, 6, 6],
             'Providence': [7, 3, 3, 6, 7]}
demands = [60, 100, 110, 50, 80]
capacities = {'Amherst': 200, 'Providence': 210}

# Create a new Gurobi model
model = gp.Model("WonderPlastics")

# Decision variables
x = {}
for plant in plants:
    for customer in customers:
        x[plant, customer] = model.addVar(name=f"x_{plant}_{customer}", vtype=GRB.CONTINUOUS, lb=0)  # Ensure non-negativity

# Objective function
model.setObjective(gp.quicksum((production_costs[plant][customer-1] + transportation_costs[plant][customer-1]) * x[plant, customer]
                                for plant in plants for customer in customers), GRB.MINIMIZE)

# Constraints
for plant in plants:
    model.addConstr(gp.quicksum(x[plant, customer] for customer in customers) <= capacities[plant], name=f"capacity_{plant}")

for customer, demand in zip(customers, demands):
    model.addConstr(gp.quicksum(x[plant, customer] for plant in plants) == demand, name=f"demand_{customer}")

# Optimize the model
model.optimize()

# Print the optimal solution
if model.status == GRB.OPTIMAL:
    print("Optimal solution found!")
    for plant in plants:
        for customer in customers:
            print(f"Units from {plant} to customer {customer}: {x[plant, customer].x}")
    print(f"Total cost: ${model.objVal}")
else:
    print("Optimization did not converge.")


Gurobi Optimizer version 11.0.0 build v11.0.0rc2 (linux64 - "Ubuntu 22.04.3 LTS")

CPU model: Intel(R) Xeon(R) CPU @ 2.20GHz, instruction set [SSE2|AVX|AVX2]
Thread count: 1 physical cores, 2 logical processors, using up to 2 threads

Optimize a model with 7 rows, 10 columns and 20 nonzeros
Model fingerprint: 0x975f3941
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [4e+00, 1e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [5e+01, 2e+02]
Presolve removed 6 rows and 7 columns
Presolve time: 0.01s
Presolved: 1 rows, 4 columns, 4 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    3.0200000e+03   7.500000e+00   0.000000e+00      0s
       1    3.1200000e+03   0.000000e+00   0.000000e+00      0s

Solved in 1 iterations and 0.02 seconds (0.00 work units)
Optimal objective  3.120000000e+03
Optimal solution found!
Units from Amherst to customer 1: 60.0
Units from Amherst to customer 2: 0.0
Units from Amherst to customer

In [ ]:
import gurobipy as gp
from gurobipy import GRB

# Data
plants = ['Amherst', 'Providence']
customers = [1, 2, 3, 4, 5]
production_costs = {'Amherst': [5, 3, 2, 6, 6],
                    'Providence': [7, 3, 3, 6, 7]}
transportation_costs = {'Amherst': [2, 3, 2, 6, 6],
                        'Providence': [7, 3, 3, 6, 7]}
demands = [60, 100, 110, 50, 80]
capacities = {'Amherst': 200, 'Providence': 210}

# Create a new Gurobi model
model = gp.Model("WonderPlastics")

# Decision variables
x = {}
for plant in plants:
    for customer in customers:
        x[plant, customer] = model.addVar(name=f"x_{plant}_{customer}", vtype=GRB.CONTINUOUS)

# Objective function
model.setObjective(gp.quicksum((production_costs[plant][customer-1] + transportation_costs[plant][customer-1]) * x[plant, customer]
                                for plant in plants for customer in customers), GRB.MINIMIZE)

# Constraints
for plant in plants:
    model.addConstr(gp.quicksum(x[plant, customer] for customer in customers) <= capacities[plant], name=f"capacity_{plant}")

for customer, demand in zip(customers, demands):
    model.addConstr(gp.quicksum(x[plant, customer] for plant in plants) == demand, name=f"demand_{customer}")

# Optimize the model
model.optimize()

# Print the optimal solution
if model.status == GRB.OPTIMAL:
    print("Optimal solution found!")
    for plant in plants:
        for customer in customers:
            print(f"Units from {plant} to customer {customer}: {x[plant, customer].x}")
    print(f"Total cost: ${model.objVal}")
else:
    print("Optimization did not converge.")


Gurobi Optimizer version 11.0.0 build v11.0.0rc2 (linux64 - "Ubuntu 22.04.3 LTS")

CPU model: Intel(R) Xeon(R) CPU @ 2.20GHz, instruction set [SSE2|AVX|AVX2]
Thread count: 1 physical cores, 2 logical processors, using up to 2 threads

Optimize a model with 7 rows, 10 columns and 20 nonzeros
Model fingerprint: 0x975f3941
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [4e+00, 1e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [5e+01, 2e+02]
Presolve removed 6 rows and 7 columns
Presolve time: 0.01s
Presolved: 1 rows, 4 columns, 4 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    3.0200000e+03   7.500000e+00   0.000000e+00      0s
       1    3.1200000e+03   0.000000e+00   0.000000e+00      0s

Solved in 1 iterations and 0.02 seconds (0.00 work units)
Optimal objective  3.120000000e+03
Optimal solution found!
Units from Amherst to customer 1: 60.0
Units from Amherst to customer 2: 0.0
Units from Amherst to customer

Units from Amherst to customer 1: 60.0

Units from Amherst to customer 2: 0.0

Units from Amherst to customer 3: 60.0

Units from Amherst to customer 4: 0.0

Units from Amherst to customer 5: 80.0

Units from Providence to customer 1: 0.0

Units from Providence to customer 2: 100.0

Units from Providence to customer 3: 50.0

Units from Providence to customer 4: 50.0

Units from Providence to customer 5: 0.0
Total cost: $3120.0